File: ui/desktop/src/schedule.ts
import {
  listSchedules as apiListSchedules,
  createSchedule as apiCreateSchedule,
  deleteSchedule as apiDeleteSchedule,
  pauseSchedule as apiPauseSchedule,
  unpauseSchedule as apiUnpauseSchedule,
  updateSchedule as apiUpdateSchedule,
  sessionsHandler as apiGetScheduleSessions,
  runNowHandler as apiRunScheduleNow,
} from './api';

export interface ScheduledJob {
  id: string;
  source: string;
  cron: string;
  last_run?: string | null;
  currently_running?: boolean;
  paused?: boolean;
}

export interface ScheduleSession {
  id: string;
  name: string;
  createdAt: string; // ISO 8601 date string
  workingDir: string;
  scheduleId: string;
  messageCount: number;
  totalTokens: number;
  inputTokens: number;
  outputTokens: number;
  accumulatedTotalTokens: number;
  accumulatedInputTokens: number;
  accumulatedOutputTokens: number;
}

export async function listSchedules(): Promise<ScheduledJob[]> {
  try {
    const response = await apiListSchedules<true>();
    if (response && response.data && Array.isArray(response.data.jobs)) {
      return response.data.jobs as ScheduledJob[];
    }
    console.error('Unexpected response format from apiListSchedules', response);
    throw new Error('Failed to list schedules: Unexpected response format');
  } catch (error) {
    console.error('Error listing schedules:', error);
    throw error;
  }
}

export async function createSchedule(request: {
  id: string;
  recipe_source: string;
  cron: string;
}): Promise<ScheduledJob> {
  try {
    const response = await apiCreateSchedule<true>({ body: request });
    if (response && response.data) {
      return response.data as ScheduledJob;
    }
    console.error('Unexpected response format from apiCreateSchedule', response);
    throw new Error('Failed to create schedule: Unexpected response format');
  } catch (error) {
    console.error('Error creating schedule:', error);
    throw error;
  }
}

export async function deleteSchedule(id: string): Promise<void> {
  try {
    await apiDeleteSchedule<true>({ path: { id } });
  } catch (error) {
    console.error(`Error deleting schedule ${id}:`, error);
    throw error;
  }
}

export async function getScheduleSessions(
  scheduleId: string,
  limit?: number
): Promise<ScheduleSession[]> {
  try {
    const response = await apiGetScheduleSessions<true>({
      path: { id: scheduleId },
      query: { limit },
    });

    if (response && response.data) {
      return response.data as ScheduleSession[];
    }
    console.error('Unexpected response format from apiGetScheduleSessions', response);
    throw new Error('Failed to get schedule sessions: Unexpected response format');
  } catch (error) {
    console.error(`Error fetching sessions for schedule ${scheduleId}:`, error);
    throw error;
  }
}

export async function runScheduleNow(scheduleId: string): Promise<string> {
  try {
    const response = await apiRunScheduleNow<true>({
      path: { id: scheduleId },
    });

    if (response && response.data && response.data.session_id) {
      return response.data.session_id;
    }
    console.error('Unexpected response format from apiRunScheduleNow', response);
    throw new Error('Failed to run schedule now: Unexpected response format');
  } catch (error) {
    console.error(`Error running schedule ${scheduleId} now:`, error);
    throw error;
  }
}

export async function pauseSchedule(scheduleId: string): Promise<void> {
  try {
    await apiPauseSchedule<true>({
      path: { id: scheduleId },
    });
  } catch (error) {
    console.error(`Error pausing schedule ${scheduleId}:`, error);
    throw error;
  }
}

export async function unpauseSchedule(scheduleId: string): Promise<void> {
  try {
    await apiUnpauseSchedule<true>({
      path: { id: scheduleId },
    });
  } catch (error) {
    console.error(`Error unpausing schedule ${scheduleId}:`, error);
    throw error;
  }
}

export async function updateSchedule(scheduleId: string, cron: string): Promise<ScheduledJob> {
  try {
    const response = await apiUpdateSchedule<true>({
      path: { id: scheduleId },
      body: { cron },
    });

    if (response && response.data) {
      return response.data as ScheduledJob;
    }
    console.error('Unexpected response format from apiUpdateSchedule', response);
    throw new Error('Failed to update schedule: Unexpected response format');
  } catch (error) {
    console.error(`Error updating schedule ${scheduleId}:`, error);
    throw error;
  }
}

File: crates/goose-server/src/routes/schedule.rs
use std::sync::Arc;

use axum::{
    extract::{Path, Query, State}, // Added Query
    http::StatusCode,
    routing::{delete, get, post},
    Json,
    Router,
};
use serde::{Deserialize, Serialize};

// Added for parsing session_name to created_at
use chrono::NaiveDateTime;

use crate::state::AppState;
use goose::scheduler::ScheduledJob;

#[derive(Deserialize)]
struct CreateScheduleRequest {
    id: String,
    recipe_source: String,
    cron: String,
}

#[derive(Serialize)]
struct ListSchedulesResponse {
    jobs: Vec<ScheduledJob>,
}

// Response for the run_now endpoint
#[derive(Serialize)]
struct RunNowResponse {
    session_id: String,
}

// Query parameters for the sessions endpoint
#[derive(Deserialize)]
struct SessionsQuery {
    #[serde(default = "default_limit")]
    limit: u32,
}

fn default_limit() -> u32 {
    50 // Default limit for sessions listed
}

// Struct for the frontend session list
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct SessionDisplayInfo {
    id: String,          // Derived from session_name (filename)
    name: String,        // From metadata.description
    created_at: String,  // Derived from session_name, in ISO 8601 format
    working_dir: String, // from metadata.working_dir (as String)
    schedule_id: Option<String>,
    message_count: usize,
    total_tokens: Option<i32>,
    input_tokens: Option<i32>,
    output_tokens: Option<i32>,
    accumulated_total_tokens: Option<i32>,
    accumulated_input_tokens: Option<i32>,
    accumulated_output_tokens: Option<i32>,
}

fn parse_session_name_to_iso(session_name: &str) -> String {
    NaiveDateTime::parse_from_str(session_name, "%Y%m%d_%H%M%S")
        .map(|dt| dt.and_utc().to_rfc3339())
        .unwrap_or_else(|_| String::new()) // Fallback to empty string if parsing fails
}

#[axum::debug_handler]
async fn create_schedule(
    State(state): State<Arc<AppState>>,
    Json(req): Json<CreateScheduleRequest>,
) -> Result<Json<ScheduledJob>, StatusCode> {
    let scheduler = state
        .scheduler()
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;
    let job = ScheduledJob {
        id: req.id,
        source: req.recipe_source,
        cron: req.cron,
        last_run: None,
    };
    scheduler
        .add_scheduled_job(job.clone())
        .await
        .map_err(|e| {
            eprintln!("Error creating schedule: {:?}", e); // Log error
            StatusCode::INTERNAL_SERVER_ERROR
        })?;
    Ok(Json(job))
}

#[axum::debug_handler]
async fn list_schedules(
    State(state): State<Arc<AppState>>,
) -> Result<Json<ListSchedulesResponse>, StatusCode> {
    let scheduler = state
        .scheduler()
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;
    let jobs = scheduler.list_scheduled_jobs().await;
    Ok(Json(ListSchedulesResponse { jobs }))
}

#[axum::debug_handler]
async fn delete_schedule(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> Result<StatusCode, StatusCode> {
    let scheduler = state
        .scheduler()
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;
    scheduler.remove_scheduled_job(&id).await.map_err(|e| {
        eprintln!("Error deleting schedule '{}': {:?}", id, e); // Log error
        match e {
            goose::scheduler::SchedulerError::JobNotFound(_) => StatusCode::NOT_FOUND,
            _ => StatusCode::INTERNAL_SERVER_ERROR,
        }
    })?;
    Ok(StatusCode::NO_CONTENT)
}

#[axum::debug_handler]
async fn run_now_handler(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> Result<Json<RunNowResponse>, StatusCode> {
    let scheduler = state
        .scheduler()
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    match scheduler.run_now(&id).await {
        Ok(session_id) => Ok(Json(RunNowResponse { session_id })),
        Err(e) => {
            eprintln!("Error running schedule '{}' now: {:?}", id, e); // Log error
            match e {
                goose::scheduler::SchedulerError::JobNotFound(_) => Err(StatusCode::NOT_FOUND),
                _ => Err(StatusCode::INTERNAL_SERVER_ERROR),
            }
        }
    }
}

#[axum::debug_handler]
async fn sessions_handler(
    State(state): State<Arc<AppState>>,
    Path(schedule_id_param): Path<String>, // Renamed to avoid confusion with session_id
    Query(query_params): Query<SessionsQuery>,
) -> Result<Json<Vec<SessionDisplayInfo>>, StatusCode> {
    // Changed return type
    let scheduler = state
        .scheduler()
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    match scheduler
        .sessions(&schedule_id_param, query_params.limit as usize)
        .await
    {
        Ok(session_tuples) => {
            // Expecting Vec<(String, goose::session::storage::SessionMetadata)>
            let display_infos: Vec<SessionDisplayInfo> = session_tuples
                .into_iter()
                .map(|(session_name, metadata)| SessionDisplayInfo {
                    id: session_name.clone(),
                    name: metadata.description, // Use description as name
                    created_at: parse_session_name_to_iso(&session_name),
                    working_dir: metadata.working_dir.to_string_lossy().into_owned(),
                    schedule_id: metadata.schedule_id, // This is the ID of the schedule itself
                    message_count: metadata.message_count,
                    total_tokens: metadata.total_tokens,
                    input_tokens: metadata.input_tokens,
                    output_tokens: metadata.output_tokens,
                    accumulated_total_tokens: metadata.accumulated_total_tokens,
                    accumulated_input_tokens: metadata.accumulated_input_tokens,
                    accumulated_output_tokens: metadata.accumulated_output_tokens,
                })
                .collect();
            Ok(Json(display_infos))
        }
        Err(e) => {
            eprintln!(
                "Error fetching sessions for schedule '{}': {:?}",
                schedule_id_param, e
            );
            Err(StatusCode::INTERNAL_SERVER_ERROR)
        }
    }
}

pub fn routes(state: Arc<AppState>) -> Router {
    Router::new()
        .route("/schedule/create", post(create_schedule))
        .route("/schedule/list", get(list_schedules))
        .route("/schedule/delete/{id}", delete(delete_schedule)) // Corrected
        .route("/schedule/{id}/run_now", post(run_now_handler)) // Corrected
        .route("/schedule/{id}/sessions", get(sessions_handler)) // Corrected
        .with_state(state)
}
